File: algorithms/js/twoPointers/trappingRainWater.js
const trapRainWater = (nums) => {
    let left = 0;
    let right = nums.length - 1;
    let leftMax = 0;
    let rightMax = 0;
    let rainCount = 0;

    while (left < right) {
        if (nums[left] < nums[right]) {
            if (nums[left] >= leftMax) {
                leftMax = nums[left];
            } else {
                rainCount += leftMax - nums[left];
            }
            left++;
        } else {
            if (nums[right] >= rightMax) {
                rightMax = nums[right];
            } else {
                rainCount += rightMax - nums[right];
            }
            right--;
        }
    }

    return rainCount;
}

File: algorithms/py/twoPointer/trappingRainWater.py
from typing import List

class Solution:
    def trap(self, heights: List[int]) -> int:
        totalRain = 0
        leftMax = 0
        rightMax = 0
        left = 0
        right = len(heights) - 1

        while left < right:
            if heights[left] < heights[right]:
                if heights[left] > leftMax:
                    leftMax = heights[left]
                else:
                    totalRain += leftMax - heights[left]
                left += 1
            else:
                if heights[right] > rightMax:
                    rightMax = heights[right]
                else:
                    totalRain += rightMax - heights[right]
                right -= 1
        
        return totalRain